design/plankton/CommunicationLayerDesign.md

```diff
@@ -36,13 +36,161 @@ The IBEX team at ISIS, for example, have expressed interest in using Plankton. N
+
+
+
+#### Environment & Timing Loop
+
+The design separates the timing loop from the `Adapter`. The while loop is built into an `Environment`, of which there is just one per simulation. We would like to be able to make updates to the `Environment` without having ot restart the simulation or renew the adapters. One possible implementation of the `Environment` might look like this:
+
+```python
+
+class Environment(object):
+
+    def __init__(self, adapter):
+        self._adapter = adapter # adapter wraps device (see above)
+        self._count = 0
+        self._timer = 0.0
+        self._delta = 0.0
+        self._user_delta = None
+        
+    def _delta(self):
+        if self._user_delta:
+            return self._user_delta
+        else:
+            return (datetime.now() - start).total_seconds()
+            
+    def set_user_delta(self, delta):
+        self._user_delta = delta
+
+    def run(self):
+
+        while True:
+            start = datetime.now()
+
+            
+            self._adapter.process(delta) # stream adapters would call asyncore.loop(0.1, count=1) as part of this
+
+            
+            count += 1
+            timer += self._delta()
+            self._delta = self._delta()
+            # ....
+```
+
+#### Environment Updates
+
+The separation of the runnable parts of the simulation (now in the `Environment`) from the `Device` and the `Adapter` parts should make updates which affect only the `Environment` easier.
+
+Here's one possibility I've been considering:
+
+`EnvironmentAdapter` sees `Environment` 
+
+```python 
+    class Environment(object):
+
+    def __init__(self, adapter, environ_adapter):
+        self._environ_adapter = environ_adapter
+        self._environ_adapter.set_adaptee(self)
+        self._adapter = adapter
+
+    def run(self):
+
+        while True:
+            # Everything still driven from a central processing loop.
+            self._environment_adapter.process()
+            
+            self._adapter.process(delta)
+
+```
+
+```python
+   class StreamEnvironmentAdapter(object):
+
+    def set_adaptee(self, environment):
+        self._environment = environment
+        
+    
+    def process(self):
+        asyncore.loop(0.1, count=1)
+        # ...
+        self._environment.set_user_delta(user_value)
+        
+```
+
+After discussion with Michael Wedel, I am now strongly leaning towards providing a simple single protocol for the `Environment Control`. While `Adapter` flexibility and variation is vital in order to support different device types. the `Environment` is internal to plankton and we should publish and control a single mechanism for this. Michael Wedel has suggested JSON over zeromq.
```

Comment:
As I said I would do earlier on this morning:

1. I've talked to Isabella via her Issue. telnet is not mandatory, but she does want to be able to set key-value pairs. It turns out they want to be able to automated test their production IOCs (Novel Idea!) with the help of the emulators, for which you actually need to have two concepts (I was trying to keep them separate). (a) The ability to modify, connect, disconnect, and modify the emulator as part of the test scenario. (b) The ability to run the emulator and get back host + port information so that you connect to it. 
1. I've had a look at the `Controllable` work that @MichaelWedel did for the live listener. Here are my thoughts.
    1. I think the `control.py` interface is pretty nice. And actually more convienient than telnet IMHO. I had originally thought that the user would have to construct the JSON by hand, which is one of the reasons I was initially not in favour of this approach.
    1. I like the auto-discovery on the server and the runtime introspection.
    1. Seems like you had to write quite a bit of code for this, so maybe finding a way to reduce the amount of internal code JSON-RPC maybe would be a good idea.
    1. It would be good if we could get some code sharing between the live data prototype and plankton. Approaches seems similar, but there should only be one `Controllable` per device. Not sure how this deviates from the live listener.
    1. It's good that you have some experience in this area. And most of the code looks like it would work for what we want to do for a manual/commandline driven process.